site/content/en/concepts/architecture/_index.md

```diff
@@ -0,0 +1,156 @@
+- kubernetes: OpenAPI read from apiserver
+
+### Resource + Controller architecture
+
+The desired state of the system should be expressed as declarative, static data structures
+rather than as executable or interpreted code (DSLs and templates are code).  
+
+Dynamic logic should be built into programs which read configuration as input, modify it
+(creating, updating or deleting resources) and emit configuration as output.
+
+It should be possible to compose multiple tools together through pipelining their input / output.
+
+#### kpt vs kubernetes:
+
+- kpt:
+  - read / write files
+  - triggered explicitly (kpt invocations)
+- kubernetes:
+  - read / write http requests
+  - triggered implicitly (watches)
+
+## IO
+
+Both kpt inputs and outputs should be recognized by Kubernetes project tools, published as
+Kubernetes examples or published by the Kubernetes apiserver.
+
+Examples:
+
+- `kubectl get -o yaml`
+- `kubectl apply -f -`
+- `github.com/kubernetes/examples`
+
+Much like Kubernetes controllers, kpt should be able to read its previous outputs and modify
+them, rather than generating them from scratch -- e.g. read a directory of
+configuration and write back to the same directory.
+
+**Note:** This requires symmetric inputs and outputs.
+
+Dynamic logic may be written using templates or DSLs -- which would not support the read-write
+workflow -- by merging the newly generated template / DSL output resources with the input
+resources.
+
+### Targets
+
+Unlike the Kubernetes control-plane, which reads and writes from the apiserver, kpt
+reads and writes from arbitrary, sources so long as they provide resource configuration.
+
+- Local files
+- Files stored in git
+- Command stdin + stdout
+- Apiserver endpoints
+
+## Configuration as Data vs Configuration as Code
+
+kpt structures configuration as data (APIs) rather than as code (templates /  DSLs).  By
+representing configuration as data, kpt promotes an extensible ecosystem of loosely coupled
+tools working against a shared data model (the Kubernetes resource model).
+
+Data may be shared by multiple tools (each reading and updating the data), whereas code is forked
+when it is modified by a new owner.
+
+Users of kpt may extend its functionality either through the built-in extension mechanisms
+(functions) or through building separate tools which follow the same resource-oriented architecture.
+
+With an "as data" model, abstractions may be implemented as APIs (in arbitrary languages)
+rather than as shared libraries.  APIs may be composed and used together independent of what
+language each API is implemented in.
+
+## Shift-Left & GitOps
+
+By shifting resource operations to be performed on local configuration rather than on remote
+cluster objects, kpt enables the ability to catch issues earlier.
+
+Cluster changes may be reviewed, approved, validated, audited and rolled back using git.
+
+### Shift-Right vs Shift-Left
+
+- Controllers => Functions
```

Comment:
this list is cryptic, maybe use some verbs

Reply:
removed